rex/joint_pd/joint_pd.py

```diff
@@ -0,0 +1,413 @@
+                logger.debug('Writing site meta data to {}'
+                             .format(meta_fpath))
+                meta.to_csv(meta_fpath, index=False)
+        elif out_fpath.endswith('.h5'):
+            with h5py.File(out_fpath) as f:
+                f.attrs['rex version'] = __version__
+                for i, n in enumerate(jpd.index.names):
+                    logger.info('')
+                    data = np.array(jpd.index.get_level_values(i))
+                    dset = '{}-bins'.format(n)
+                    logger.debug('Writing {}'.format(dset))
+                    f.create_dataset(dset, shape=data.shape, dtype=data.dtype,
+                                     data=data)
+
+                logger.debug('Writing joint probability density values to jpd')
+                data = jpd.values
+                f.create_dataset('jpd', shape=data.shape, dtype=data.dtype,
+                                 data=data)
+
+                logger.debug('Writing site meta data to meta')
+                meta = to_records_array(meta)
+                f.create_dataset('meta', shape=meta.shape, dtype=meta.dtype,
+                                 data=meta)
+        else:
+            msg = ("Cannot save joint probability distribution, expecting "
+                   ".csv or .h5 path, but got: {}".format(out_fpath))
+            logger.error(msg)
+            raise OSError(msg)
+
+    @classmethod
+    def run(cls, res_h5, dset1, dset2, bins1, bins2,
+            sites=None, res_cls=Resource, hsds=False,
+            max_workers=None, chunks_per_worker=5, out_fpath=None):
+        """
+        Compute joint probability distribution between given datasets using
+        given bins
+
+        Parameters
+        ----------
+        res_h5 : str
+            Path to resource h5 file(s)
+        dset1 : str
+            Dataset 1 to generate joint probability distribution for
+        dset2 : str
+            Dataset 2 to generate joint probabilty distribution for
+        bin1 : tuple
+            (start, stop, step) for dataset 1 bins
+        bin2 : tuple
+            (start, stop, step) for dataset 2 bins
+        sites : list | slice, optional
+            Subset of sites to extract, by default None or all sites
+        res_cls : Class, optional
+            Resource class to use to access res_h5, by default Resource
+        hsds : bool, optional
+            Boolean flag to use h5pyd to handle .h5 'files' hosted on AWS
+            behind HSDS, by default False
+        max_workers : None | int, optional
+            Number of workers to use, if 1 run in serial, if None use all
+            available cores, by default None
+        chunks_per_worker : int, optional
+            Number of chunks to extract on each worker, by default 5
+        out_fpath : str, optional
+            .csv, or .h5 file path to save joint probability
+            distribution to
+
+        Returns
+        -------
+        out : pandas.DataFrame
```

Comment:
How in the world are you putting 2D histograms for multiple sites into a single dataframe?

Reply:
Here is a wind rose example:
/scratch/mrossol/NOWRDC_wind_roses/Mid_Atlantic_2019_wind_rose-119m.csv

The format is
dset 1 bin, dset 2 bin, site 0 density, site 1 density, ...

Reply:
Ah I see. I think there should be an entry point for single-site that returns the 2D array (or df) that you would expect of a 2D histogram. But thats just a minor thing I would expect.

Reply:
The problem is that each side produces a 2d array, to fit it into the dataframe I have to flatten it:
https://github.com/NREL/rex/blob/f2a928043ff5ad74cb3cb39bef9c024d55c7d1eb/rex/joint_pd/joint_pd.py#L171-L172

I was going to add a plotting utility that will let you give the dataframe and pick a site and it will produce a heatmap

Reply:
Yeah gotcha but people might also want to use the JointDist class as a utility for single arrays and they would expect to get a 2D array out. I guess they could just use numpy directly but it would be nice to have plotting and eventually sampling utilities (I've made this for a different project).

Reply:
The whole point of the class and ask from both the Wave guys and for the WindRose was the need to efficiently run the JPD for all sites over 1 or many years.

I say just run np.hist2d if you want the distribution for a single site. I could re-name, expose, and tweak the logic here to allow for running a single site:
https://github.com/NREL/rex/blob/f2a928043ff5ad74cb3cb39bef9c024d55c7d1eb/rex/joint_pd/joint_pd.py#L119-L174

Reply:
I just think there is a lot of ancillary functionality that makes joint distributions very useful that are not present in the numpy method. But I agree that this class can definitely just focus on joint dists for spatiotemporal data and thats totally appropriate.